### PR TITLE
[Serializer] ensure ChainEncoder/ChainDecoder properly consider the context

### LINKED ISSUE #38270: [Serializer] ChainEncoder does not honor context support check if called repeatedly
`ChainEncoder#getEncoder` method reads:

```php
        if (isset($this->encoderByFormat[$format])
            && isset($this->encoders[$this->encoderByFormat[$format]])
        ) {
            return $this->encoders[$this->encoderByFormat[$format]];
        }

        foreach ($this->encoders as $i => $encoder) {
            if ($encoder->supportsEncoding($format, $context)) {
                $this->encoderByFormat[$format] = $i;

                return $encoder;
            }
        }
```

If that method is called more than once with different `$context` arguments the wrong encoder could be returned because `$encoder->supportsEncoding($format, $context)` is not called again.

Hence, I disabled the first check and `return` statement and things worked like a charm.

### PR DESCRIPTION
**Pull request by Guite**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #38270 
| License       | MIT
| Doc PR        | -

This PR removes the "caching" of encoder/decoder instances in a chain so they are re-evaluated for each single call.

The corresponding tests are extended to perform multiple calls with different context arrays. Also they now check not only if encoding/decoding is supported, but also that the responsible encoder/decoder instance is the correct (expected) one.

/cc @xabbuh

### CODE REVIEW

**Comment by dunglas** [comment]
We must be very careful when touching this part. Removing such cache can have a dramatic performance impact, as it's on the hot path. We need to profile and to double-check that it doesn't make complex apps very slow. We'll probably have to do something like this: https://github.com/symfony/symfony/pull/27105

cc @nicolas-grekas.

File: src/Symfony/Component/Serializer/Encoder/ChainDecoder.php [hunk 1]
```diff
@@ -59,18 +58,10 @@ public function supportsDecoding($format, array $context = []): bool
      *
      * @throws RuntimeException if no decoder is found
      */
-    private function getDecoder(string $format, array $context): DecoderInterface
+    public function getDecoder(string $format, array $context): DecoderInterface
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Encoder/ChainDecoder.php]: Why this change? This will make this code harder to maintain. I think that this method should remain private.

Comment (Guite) [reply src/Symfony/Component/Serializer/Encoder/ChainDecoder.php]: Because the tests now retrieve/expect a specific encoder/decoder instance.

Comment (dunglas) [reply src/Symfony/Component/Serializer/Encoder/ChainDecoder.php]: We cannot modify our public API just for a test. This should be tested in another way.
Here you can just register a test encoder returning a known result  ("magic") when properly called.

**Comment by dunglas** [comment]
I think that we should just replicate the logic introduced for normalizers in #27105. Cache when not dynamic, and don't cache when it is.

File: src/Symfony/Component/Serializer/Encoder/ChainDecoder.php (hunk 1 above)
Comment (Guite) [reply src/Symfony/Component/Serializer/Encoder/ChainDecoder.php]: Okay. I can change that - if the general change of this PR is fine for you.

**Comment by lyrixx** [comment]
Another implementation of the chain could be added, it would be simpler IMHO, since it's an edge case

**Comment by Guite** [comment]
> Cache when not dynamic, and don't cache when it is.

Dynamic would be a context that is not just an empty array?

**Comment by Guite** [comment]
I readded caching arrays but utilize them only if no dynamic context is given.

File: src/Symfony/Component/Serializer/Encoder/ChainDecoder.php [hunk 2]
```diff
@@ -59,17 +59,23 @@ public function supportsDecoding($format, array $context = []): bool
...
     {
-        if (isset($this->decoderByFormat[$format])
+        $hasContext = 0 < \count($context);
+        if (
+            true !== $hasContext
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Encoder/ChainDecoder.php]: please remove comparisons to booleans everywhere

Comment (Guite) [reply src/Symfony/Component/Serializer/Encoder/ChainDecoder.php]: @nicolas-grekas  you mean replace `true === $foo` by `$foo` and `true !== $foo` by `!$foo`?

Comment (stof) [reply src/Symfony/Component/Serializer/Encoder/ChainDecoder.php]: yes

File: src/Symfony/Component/Serializer/Encoder/ChainDecoder.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Serializer/Encoder/ChainDecoder.php]: It's 100% sure that we won't make these public for tests, no need to wait on that point

**Comment by nicolas-grekas** [comment]
I propose checking for `ContextAwareEncoreInterface` instead, see #47150
WDYT? Can you update the PR accordingly?

**Comment by nicolas-grekas** [comment]
Closing in favor of #47304
